lib/config.js

```diff
@@ -209,109 +149,313 @@ function Config(options) {
-    let config,
-        userConfig;
+Config.prototype.loadConfig = function(configToLoad) {
+    if (typeof configToLoad !== "string") {
+        return loadConfig(configToLoad);
+    }
 
-    debug("Constructing config for " + (filePath ? filePath : "text"));
+    const filePath = configToLoad;
+    let config = this.cache[filePath];
 
-    config = this.cache[directory];
+    if (!config) {
+        config = this.cache[filePath] = loadConfig(filePath);
+    }
+    return config;
+};
 
-    if (config) {
-        debug("Using config from cache");
-        return config;
+/**
+ * Get personal config object from user's home directory
+ * @returns {Object} the personal config object (null if there is no personal config)
+ * @private
+ */
+Config.prototype.getPersonalConfig = function() {
+    if (typeof this.personalConfig === "undefined") {
+        let config;
+
+        if (PERSONAL_CONFIG_DIR) {
+            const filename = ConfigFile.getFilenameForDirectory(PERSONAL_CONFIG_DIR);
+
+            if (filename) {
+                debug("Using personal config");
+                config = this.loadConfig(filename);
+            }
+        }
+        this.personalConfig = config || null;
     }
 
-    // Step 1: Determine user-specified config from .eslintrc.* and package.json files
+    return this.personalConfig;
+};
+
+/**
+ * Build a config hierarchy including the base config (conf/eslint.json), the
+ * environments config (conf/environments.js) and eventually the user config.
+ * @param {string} directory a file in whose directory we start looking for a local config
+ * @returns {Object[]} The config objects
+ * @private
+ */
+Config.prototype.getConfigHierarchy = function(directory) {
+    let configs;
+
+    debug("Constructing config file hierarchy for " + directory);
+
+    // Step 1: Always include baseConfig
+    configs = [ this.baseConfig ];
+
+    // Step 2: Add user-specified config from .eslintrc.* and package.json files
     if (this.useEslintrc) {
         debug("Using .eslintrc and package.json files");
-        userConfig = getLocalConfig(this, directory);
+        configs = configs.concat(this.getLocalConfigHierarchy(directory));
     } else {
         debug("Not using .eslintrc or package.json files");
-        userConfig = {};
     }
 
-    // Step 2: Create a copy of the baseConfig
-    config = ConfigOps.merge({}, this.baseConfig);
+    // Step 3: Merge in command line config file
+    if (this.useSpecificConfig) {
+        debug("Using command line config file");
+        configs.push(this.useSpecificConfig);
+    }
 
-    // Step 3: Merge in the user-specified configuration from .eslintrc and package.json
-    config = ConfigOps.merge(config, userConfig);
+    return configs;
+};
 
-    // Step 4: Merge in command line config file
-    if (this.useSpecificConfig) {
-        debug("Merging command line config file");
+/**
+ * Get the local config hierarchy for a given directory.
```

Comment:
Similar to previous note about adding comments, it's unclear what "local" means here. Can you be more specific in this comment?